src/cpp/types/fragile_ptr.hpp

```diff
@@ -0,0 +1,311 @@
+ * These shared_ptr has been replaced with fragile_ptr so code does not have to change everywhere.
+ */
+template <typename T>
+class fragile_ptr
+{
+
+public:
+
+    /////////////////////
+    // CONSTRUCTORS
+
+    //! Default constructor to nullptr
+    fragile_ptr() noexcept = default;
+
+    //! Default constructors and operator= from other \c fragile_ptr ( \c weak_ptr are copiable and moveable).
+    fragile_ptr(const fragile_ptr<T>& copy_other) noexcept = default;
+    fragile_ptr(fragile_ptr<T>&& move_other) noexcept = default;
+    fragile_ptr& operator=(const fragile_ptr<T>& copy_other) noexcept = default;
+    fragile_ptr& operator=(fragile_ptr<T>&& move_other) noexcept = default;
+
+    //! Default destructor
+    ~fragile_ptr() = default;
+
+    /////////////////////
+    // CONSTRUCTORS FROM SHARED PTR
+
+    /**
+     * @brief Construct a new fragile ptr object from a shared one
+     *
+     * It initializes the internal \c weak_ptr with the reference of the shared one.
+     *
+     * @param shared_reference \c shared_ptr to the reference to point from this.
+     */
+    fragile_ptr(const std::shared_ptr<T>& shared_reference) noexcept
+        : reference_(shared_reference)
+    {
+        // Do nothing
+    }
+
+    //! Same as calling basic constructor.
+    fragile_ptr(std::nullptr_t) noexcept
+    {
+        // Do nothing
+    }
+
+    /**
+     * @brief Assign the fragile ptr object from a shared one
+     *
+     * It initializes the internal \c weak_ptr with the reference of the shared one.
+     *
+     * @param shared_reference \c shared_ptr to the reference to point from this.
+     */
+    fragile_ptr& operator=(const std::shared_ptr<T>& shared_reference) noexcept
+    {
+        this->reference_ = shared_reference;
+        return *this;
+    }
+
+    /////////////////////
+    // OPERATOR METHODS
+
+    /**
+     * @brief Compare this ptr reference with a shared one.
+     *
+     * @param other \c shared_ptr to compare with.
+     *
+     * @todo use if_enabled or somehow limit this cast to only those U that are coherent.
+     */
+    template <typename U>
+    bool operator==(const std::shared_ptr<U>& other) const noexcept
+    {
+        static_assert(std::is_base_of<U, T>::value, "U must inherit from T");
+        return this->lock() == other;
+    }
+
+    /////////////////////
+    // CAST METHODS
+
+    /**
+     * @brief Cast this object to a \c shared_ptr object of type \c U .
+     *
+     * @todo use if_enabled or somehow limit this cast to only those U that are coherent.
+     */
+    template <typename U>
+    operator std::shared_ptr<U> () const
+    {
+        auto cast_result = std::dynamic_pointer_cast<U>(this->lock());
+        if (!cast_result)
+        {
+            // TODO: show in error message the name of the types that are being casted
+            throw Inconsistency("Trying to cast to a not valid object.");
+        }
+        else
+        {
+            return cast_result;
+        }
```

Comment:
If we only need to cast to a shared_ptr of a base class, then this method could just be:

```suggestion
        static_assert(std::is_base_of<U, T>::value, "U must inherit from T");
        return safety_lock_();
```

Reply:
I rather use lock(), as doing this cast in a normal shared_ptr requires to check the result from the user side.